File: UClient/Assets/UFramework/Scripts/Service/ManagerService.cs
/*
 * @Author: fasthro
 * @Date: 2020-12-11 13:38:20
 * @Description: manager service
 */
using LuaInterface;

namespace UFramework
{
    public class ManagerService : BaseService
    {
        public ManagerContainer container { get; set; }

        private BaseManager[] _allManagers;

        protected override void OnInitialize()
        {
            container = new ManagerContainer();

            RegisterManager(new LuaManager());
            RegisterManager(new NetManager());
            RegisterManager(new ResManager());
        }

        public void RegisterManager(BaseManager manager, bool overwriteExisting = true)
        {
            container.RegisterManager(manager, overwriteExisting);

            _allManagers = container.GetAllManagers();
        }

        [NoToLua]
        public T GetManager<T>() where T : BaseManager
        {
            return container.GetManager<T>();
        }

        public BaseManager GetManager(string managerName)
        {
            return container.GetManager(managerName);
        }

        [NoToLua]
        protected override void OnUpdate(float deltaTime)
        {
            for (int i = 0; i < _allManagers.Length; i++)
                _allManagers[i].Update(deltaTime);
        }

        [NoToLua]
        protected override void OnLateUpdate()
        {
            for (int i = 0; i < _allManagers.Length; i++)
                _allManagers[i].LateUpdate();
        }

        [NoToLua]
        protected override void OnFixedUpdate()
        {
            for (int i = 0; i < _allManagers.Length; i++)
                _allManagers[i].FixedUpdate();
        }

        [NoToLua]
        protected override void OnDispose()
        {
            for (int i = 0; i < _allManagers.Length; i++)
                _allManagers[i].Dispose();
        }

        [NoToLua]
        protected override void OnApplicationQuit()
        {
            for (int i = 0; i < _allManagers.Length; i++)
                _allManagers[i].ApplicationQuit();
        }
    }
}

File: UClient/Assets/Scripts/Lua/Service/ManagerService.lua
--[[
Author: fasthro
Date: 2020-11-11 16:04:27
Description: manager service
--]]
require("Manager.BaseManager")
require("Manager.PanelManager")
require("Manager.CtrlManager")
require("Manager.NetManager")
require("Manager.ResManager")
require("Manager.EventManager")

local ManagerService =
    typesys.def.ManagerService {
    __super = typesys.BaseService,
    _managers = typesys.map
}

function ManagerService:initialize()
    self._managers = typesys.new(typesys.map, type(""), typesys.BaseManager)

    _G.PanelManager = self:_addManager(typesys.PanelManager)
    _G.CtrlManager = self:_addManager(typesys.CtrlManager)
    _G.EventManager = self:_addManager(typesys.EventManager)
    _G.NetManager = self:_addManager(typesys.NetManager, self._ext:GetManager("NetManager"))
    _G.ResManager = self:_addManager(typesys.ResManager, self._ext:GetManager("ResManager"))
end

function ManagerService:_addManager(name, ext)
    local manager, _name = __newdtn(name, ext)
    manager:initialize()
    self._managers:set(_name, manager)
    return manager
end

function ManagerService:getManager(name)
    local _name, _dt = __parsedtn(name)
    return self._managers:get(_name)
end

function ManagerService:removeManager(name)
    local _name, _dt = __parsedtn(name)
    self._managers:set(_name, nil)
end

return ManagerService
